keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md

```diff
@@ -0,0 +1,260 @@
+  - CCMs and overall scope for moving out of tree is in [removing-in-tree-providers](keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+
+## Proposal
+
+We will add a [build constraints](https://golang.org/pkg/go/build/#hdr-Build_Constraints)
+to the cloud provider code for a pseudo "build tag" specifying not to include
+any in-tree provider code. This will allow compiling the binaries as normal today
+and simulating the removal of this code by specifying the tag at build time and
+triggering the build constraints on the files in these packages.
+
+Some small adjustments may be necessary to the code base to ensure that the
+other packages can build without depending on these packages.
+
+A prototype is available in [kubernetes/kubernetes#80353](https://github.com/kubernetes/kubernetes/pull/80353).
+
+To ensure that this continues to function we will add CI building in this mode,
+and CI running end to end tests against it (see the test plan).
+
+### User Stories [optional]
+
+#### Story 1
+
+As an out of tree cloud provider implementer, I want to develop and test against
+Kubernetes without the in tree providers.
+
+Kubernetes out of tree cloud provider developers will be able to build Kubernetes
+in this mode and build & test their cloud-controller-manager implementations and
+associated tooling against this build in preparation for the actual hard removal
+of the in-tree providers.
+
+#### Story 2
+
+As a developer working to replace an in-tree provider with an out-of-tree provider,
+I am attempting to validate that I work with KAS/KCM/Kubelet which do not have 
+(my) in-tree cloud-provider compiled in and have successfully migrated all the 
+functionality I need to CCM/CSI/... Using this build ensures the relevant 
+functionality is not in KAS/KCM/Kubelet. It also allows me to work with 
+smaller binaries.
+
+#### Story 3
+
+As a [kind](https://github.com/kubernetes-sigs/kind) developer / user I want to 
+use Kubernetes binaries without cloud providers for local clusters.
+
+Developers and users will be able to build local clusters leveraging this mode
+to not pay for cloud providers they are unable to use.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This is implemented using a synthetic `nolegacyproviders` tag in go build
+constraints on the relevant sources. If `GOFLAGS=-tags=nolegacyproviders` is set
+then the legacy cloud provider pacakges will be excluded from the build.
+
+In order to make this work the following additional changes are made:
+
+- Packages that we fully exclude (the legacy provider packages) _must_ contain
+a `doc.go` file or any other file that does NOT contain any code or build 
+constraints. Go will not allow "building" a package without any files passing
+the constraints, however it will happily build a package with no actual methods
+/ variables / ...
+
+- A few locations in the code do not properly use the cloud provider interface,
```

Comment:
what does "properly use" mean in this context

Reply:
they don't use it. my understanding is we have a "cloud provider interface" and a location where we load all of the provider via underscore import, but some code imports the legacy cloud provider packages directly.

Reply:
I added a clarification here.

Reply:
So we have things line credential providers (https://github.com/kubernetes/kubernetes/issues/74379) which contain cloud provider specific code but it does not go through any sort of pluggable code to add that.